### DISCUSSION TITLE
Undercurl not working anywhere

### DISCUSSION THREAD

**Discussion by simonmandlik** [neovim/neovim#24603] [Q&A] [answered]
I cannot get undercurl to work. I tested all of the "mainstream" terminal emulators — `iTerm2`, `kitty`, and `alacritty`, with `zsh`. Undercurl works in all terminals, but not in `neovim`.

iTerm2 shell:
<img width="370" alt="image" src="https://github.com/neovim/neovim/assets/16707112/a4fa396e-aa84-4d62-a4bf-4f3b9d26adae">

iTerm2 `neovim`:
<img width="535" alt="image" src="https://github.com/neovim/neovim/assets/16707112/1530aa71-130a-4d34-bf93-89c5d5d8c191">

kitty shell:
<img width="507" alt="image" src="https://github.com/neovim/neovim/assets/16707112/45dcf549-acc0-46e3-bcec-361a485198b3">

kitty `neovim`:
<img width="732" alt="image" src="https://github.com/neovim/neovim/assets/16707112/a9b8dc05-8868-44ee-83fa-857774616fd5">

alacrity shell:
<img width="470" alt="image" src="https://github.com/neovim/neovim/assets/16707112/44fcf452-567f-4b3f-b579-05c66988de34">

alacritty `neovim`:
<img width="679" alt="image" src="https://github.com/neovim/neovim/assets/16707112/43d8ce7f-8d2b-4cd4-845b-9af6978f8c33">

I'm on `NVIM v0.9.1 Build type: Release LuaJIT 2.1.0-beta3` and `macOS 13.4.1`.

Thanks for any help!

**Comment by clason** [comment]
Try actually setting your highlight groups correctly: if you have `termguicolors` set, you should set `guifg/guibg` and `gui`; otherwise it's `cterm`; `:h highlight-args`. 

Here's mine in `iTerm2`:
<img width="643" alt="Screenshot 2023-08-07 at 19 50 37" src="https://github.com/neovim/neovim/assets/2361214/bdd7bd30-b022-43b6-849f-342c947f94e2">

**Reply by simonmandlik** [reply]
Thanks for the answer!

The example is all defaults, with `termguicolors` off. If I `set termguicolors`, I get this:

<img width="437" alt="image" src="https://github.com/neovim/neovim/assets/16707112/d12ac7f4-f501-4828-a58d-d326ac83af68">

And your specific example

<img width="808" alt="image" src="https://github.com/neovim/neovim/assets/16707112/09f8b905-bf33-4d7a-864f-25950ed4cad2">

**Reply by clason** [reply]
Looks like your terminal doesn't support undercurl, or your TERM is wrong.

**Reply by simonmandlik** [reply]
But it does! All of the three I tested support undercurl, see the screenshots. `TERM` is set, I believe, properly as well

**Reply by clason** [reply]
"believe" gets you nowhere here. As I showed, it works, and in one of the settings you listed. That indicates misconfiguration, and wrong TERM is the culprit nine times out of ten.

**Reply by simonmandlik** [reply]
I have shared the `TERM` I'm using. For iTerm2, its' `xterm-256color`. Which one do you use?

**Reply by clason** [reply]
The default. I am not setting anything (which is a mistake, 90% of the time).

**Reply by clason** [reply]
To be clear, those screenshots are with `nvim --clean`, right?

**Reply by simonmandlik** [reply]
I'm not setting the value anywhere either. There's a `Report terminal type:` setting in `iTerm2`, which I haven't changed either, and is set to  `xterm-256color`.

Yes, this is `nvim --clean`

**Reply by clason** [reply]
Then I don't know what to tell you beyond "it works for me".

**Reply by simonmandlik** [reply]
I observed the same problem with `vim` and the solution is here: https://github.com/vim/vim/issues/12744

But `neovim` does not use the `t` codes, so it is not applicable

**Reply by clason** [reply]
Yes, this should work out of the box in Neovim (and does for most people).

**Reply by simonmandlik** [reply]
I understand, thanks for your time anyway 👍

**Reply by clason** [reply, marked as answer]
Do make sure (since it's not quite clear from your comment) to test my snippet with
1. iterm2 (latest version, beta10 IIRC)
2. `termguicolors` set.

**Reply by simonmandlik** [reply]
I got it working in the beta version! (`3.5.0beta10`)

<img width="421" alt="image" src="https://github.com/neovim/neovim/assets/16707112/ed1d75ba-d10d-4fce-b33c-3192e03b8232">

The current stable version from January (`3.4.19`) is not working
